File: algorithm/All-In-One-Java/문제/week3/src-js/Problem3.js
/**
 * Problem 3: 가장 먼 노드
 * 난이도: 🔴 Lv.3
 * 주제: BFS, 다익스트라, 그래프
 *
 * 문제: 1번 노드에서 모든 다른 노드까지의 최단 거리를 구한 후,
 *      가장 먼 거리에 있는 노드의 개수를 반환하세요.
 *
 * 입력: n = 6, edge = [[3,6],[4,3],[3,2],[1,3],[1,2],[2,4],[5,2]]
 * 출력: 3
 *
 * 핵심:
 * - BFS로 각 노드까지의 최단 거리 계산
 * - 거리를 레벨별로 관리
 * - 최댓값에 해당하는 노드 개수 세기
 */

function solution(n, edge) {
  // TODO: 풀이 작성
}

// === 테스트 ===
const tests = [
  {
    input: [
      6,
      [[3, 6], [4, 3], [3, 2], [1, 3], [1, 2], [2, 4], [5, 2]],
    ],
    expected: 3,
  },
  {
    input: [4, [[1, 2], [2, 3], [3, 4]]],
    expected: 1,
  },
];

console.log("=== Problem 3: 가장 먼 노드 ===\n");

tests.forEach((t, i) => {
  const [n, edge] = t.input;
  const result = solution(n, edge);
  const ok = result === t.expected;
  console.log(`Test ${i + 1}:`);
  console.log(`  n = ${n}`);
  console.log(`  edge = [${edge.map((e) => `[${e}]`).join(", ")}]`);
  console.log(`  결과: ${result} (기대값: ${t.expected}) ${ok ? "✓" : "✗"}\n`);
});

File: algorithm/All-In-One-Java/문제/week3/src/Problem3.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Queue;

public class Problem3 {

    static class Solution {
        public int solution(int n, int[][] edge) {
            List<List<Integer>> graph = new ArrayList<>();

            for (int i = 0; i <= n; i++) {
                graph.add(new ArrayList<>());
            }

            for (int[] e : edge) {
                int a = e[0];
                int b = e[1];

                graph.get(a).add(b);
                graph.get(b).add(a);
            }

            int[] distance = new int[n + 1];
            Arrays.fill(distance, -1);

            Queue<Integer> queue = new ArrayDeque<>();
            queue.offer(1);
            distance[1] = 0;

            while (!queue.isEmpty()) {
                int cur = queue.poll();

                for (int next : graph.get(cur)) {
                    if (distance[next] == -1) {
                        distance[next] = distance[cur] + 1;
                        queue.offer(next);
                    }
                }
            }

            int maxDistance = 0;

            for (int i = 1; i <= n; i++) {
                maxDistance = Math.max(maxDistance, distance[i]);
            }

            int answer = 0;

            for (int i = 1; i <= n; i++) {
                if (distance[i] == maxDistance) {
                    answer++;
                }
            }

            return answer;
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        System.out.println("=== Problem 3: 가장 먼 노드 ===\n");

        int[] ns = { 6, 4 };
        int[][][] edges = {
                { { 3, 6 }, { 4, 3 }, { 3, 2 }, { 1, 3 }, { 1, 2 }, { 2, 4 }, { 5, 2 } },
                { { 1, 2 }, { 1, 3 }, { 2, 4 }, { 3, 4 } }
        };
        int[] expected = { 3, 1 };

        for (int i = 0; i < ns.length; i++) {
            int n = ns[i];
            int[][] edgeArr = edges[i];
            int exp = expected[i];
            int result = sol.solution(n, edgeArr);
            String status = result == exp ? "✓" : "✗";

            System.out.println("Test " + (i + 1) + ":");
            System.out.println("  n = " + n);
            System.out.print("  edge = [");
            for (int j = 0; j < edgeArr.length; j++) {
                System.out.print(Arrays.toString(edgeArr[j]));
                if (j < edgeArr.length - 1)
                    System.out.print(", ");
            }
            System.out.println("]");
            System.out.println("  결과: " + result + " (기대값: " + exp + ") " + status);
            System.out.println();
        }
    }
}